samples/bluetooth/hci_uart/src/main.c

```diff
@@ -52,65 +58,126 @@ static K_FIFO_DEFINE(uart_tx_queue);
  */
 #define H4_DISCARD_LEN 33
 
-static int h4_read(struct device *uart, uint8_t *buf,
-		   size_t len, size_t min)
+static int h4_read(struct device *uart, uint8_t *buf, size_t len)
 {
-	int total = 0;
-
-	while (len) {
-		int rx;
-
-		rx = uart_fifo_read(uart, buf, len);
-		if (rx == 0) {
-			LOG_DBG("Got zero bytes from UART");
-			if (total < min) {
-				continue;
-			}
-			break;
-		}
+	int rx = uart_fifo_read(uart, buf, len);
 
-		LOG_DBG("read %d remaining %d", rx, len - rx);
-		len -= rx;
-		total += rx;
-		buf += rx;
-	}
+	LOG_DBG("read %d req %d", rx, len);
 
-	return total;
+	return rx;
 }
 
-static size_t h4_discard(struct device *uart, size_t len)
+static bool valid_type(uint8_t type)
 {
-	uint8_t buf[H4_DISCARD_LEN];
-
-	return uart_fifo_read(uart, buf, MIN(len, sizeof(buf)));
+	return (type == H4_CMD) | (type == H4_ACL);
 }
 
-static void h4_cmd_recv(struct net_buf *buf, int *remaining)
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static uint32_t get_len(uint8_t *hdr_buf, uint8_t type)
 {
-	struct bt_hci_cmd_hdr hdr;
-
-	/* We can ignore the return value since we pass len == min */
-	h4_read(hci_uart_dev, (void *)&hdr, sizeof(hdr), sizeof(hdr));
-
-	*remaining = hdr.param_len;
-
-	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	return (type == BT_BUF_CMD) ?
+		((struct bt_hci_cmd_hdr *)hdr_buf)->param_len :
+		sys_le16_to_cpu(((struct bt_hci_acl_hdr *)hdr_buf)->len);
+}
 
-	LOG_DBG("len %u", hdr.param_len);
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static int hdr_len(uint8_t type)
+{
+	return (type == H4_CMD) ?
+		sizeof(struct bt_hci_cmd_hdr) : sizeof(struct bt_hci_acl_hdr);
 }
 
-static void h4_acl_recv(struct net_buf *buf, int *remaining)
+static void rx_isr(void)
 {
-	struct bt_hci_acl_hdr hdr;
+	static struct net_buf *buf;
+	static int remaining;
+	static uint8_t state;
+	static uint8_t type;
+	static uint8_t hdr_buf[MAX(sizeof(struct bt_hci_cmd_hdr),
+			sizeof(struct bt_hci_acl_hdr))];
+	int read;
+
+	do {
+		switch (state) {
+		case ST_IDLE:
+			/* Get packet type */
+			read = h4_read(hci_uart_dev, &type, sizeof(type));
+			/* since we read in loop until no data is in the fifo,
+			 * it is possible that read = 0.
+			 */
+			if (read) {
+				if (valid_type(type)) {
+					/* Get expected header size and switch
+						* to receiving header.
+						*/
+					remaining = hdr_len(type);
+					state = ST_HDR;
+				} else {
+					LOG_WRN("Unknown header %d", type);
+				}
+			}
+			break;
+		case ST_HDR:
+			read = h4_read(hci_uart_dev,
+				       &hdr_buf[hdr_len(type) - remaining],
+				       remaining);
+			remaining -= read;
+			if (remaining == 0) {
+				/* Header received. Allocate buffer and get
+				 * payload length. If allocation fails leave
+				 * interrupt. Currently failed allocation is
+				 * not handled.
+				 */
+				buf = bt_buf_get_tx(BT_BUF_H4, K_NO_WAIT,
+						&type, sizeof(type));
+				__ASSERT_NO_MSG(buf);
```

Comment:
Instead of this assertion, it should be:
```
state = ST_IDLE;
return;
```
to reflect the previous implementation, since you don't want to change the behavior now.

In the current form, you'll get a null pointer dereference when assertions are disabled.

Reply:
done